### ISSUE TITLE
Large variable files have a huge impact on performance

### ISSUE THREAD

**Issue by mlaurense** [dw/mitogen#117]
Not sure whether this is Mitogen related at all, whether it is something you're looking at improving but I'm gonna go ahead and post this here. Feel free to close this if you think this isn't related, or not something you can improve.

I've got a Yaml file, used to provision a webserver via --extra-vars "@filename.yml", which is ~ 10,000 lines. This contains the configuration for the websites running on it, and the SSL certificates for certain websites. The certificates consist of two (multiline) variables, a certificate chain and an ansible-vault encrypted private key. Total size of the Yaml is about 650 KB.

The total time to run the complete playbook (no changes made, as config is the same) is about 7m36.

As a test, I've split the variable file in two parts; a SSL part with all the certificates and private keys (7849 lines, 580 KB), and a config part (2249 lines, 63 KB). The task responsible for writing the SSL certificates to the destination host was moved to a separate playbook. When running the exact same configuration, but now split into two separate playbooks using the two separate variable files, the deployment takes 8 seconds for the SSL certificates, and 5m31 to run the other playbook. This is a huge difference.

I'm guessing it has something to do with the way ansible uses the variable files, and how all variables have to be available to, for example, the ansible template module. I've also tried to move the certificates out of the domains dictionary to their own dictionary in the same Yaml file, but this makes no difference in performance.

Examples of how our Yaml is formatted: 

With certificate in domains (original):
```
domains:
  test.local:
    username: user
    enabled: true
    nginx_conf: wordpress
    php_primary: php70
    ssl: true
    ssl_parameters:
      type: custom
      certificate: |
        -----BEGIN CERTIFICATE-----
        MIIGFzCCBP+gAwIBAgISBA0r/3g+5Yp157KVS74pqITQMA0GCSqGSIb3DQEBCwUA
        [ ... ]
      certificate_key: !vault |
        $ANSIBLE_VAULT;1.1;AES256
        [ ... ]
      hsts: false
```

With certificate in a separate dict (makes no difference):
```
domains:
  test.local:
    username: user
    enabled: true
    nginx_conf: wordpress
    php_primary: php70
    ssl: true
    ssl_parameters:
      type: custom
      hsts: false
ssl_certificates:
  test.local:
    certificate: |
      -----BEGIN CERTIFICATE-----
      MIIGFzCCBP+gAwIBAgISBA0r/3g+5Yp157KVS74pqITQMA0GCSqGSIb3DQEBCwUA
      [ ... ]
    certificate_key: !vault |
      $ANSIBLE_VAULT;1.1;AES256
      [ ... ]
```

Using separate files (makes huge difference):
```
domains:
  test.local:
    username: user
    enabled: true
    nginx_conf: wordpress
    php_primary: php70
    ssl: true
    ssl_parameters:
      type: custom
      hsts: false
```
```
ssl_certificates:
  test.local:
    certificate: |
      -----BEGIN CERTIFICATE-----
      MIIGFzCCBP+gAwIBAgISBA0r/3g+5Yp157KVS74pqITQMA0GCSqGSIb3DQEBCwUA
      [ ... ]
    certificate_key: !vault |
      $ANSIBLE_VAULT;1.1;AES256
      [ ... ]
```

**Comment by dw** [comment]
Whoa that's a big file!

Can you tell me a little more about the playbook and the target machine:

* How much network latency between you and the target?
* How many steps in the playbook?
* How many super slow steps (like big apt-get jobs, etc.)
* How many with_items steps looping over that big YAML?

Finally about the YAML file itself, at this size serialization overhead is likely to be high. Would you mind running these on the file, and returning the timings:

Time parsing:

````
python -m timeit -s 'import cPickle, yaml'  'yaml.load(file("your.yml"))'
````

Time serialization:

````
python -m timeit -s 'import cPickle, yaml; y = yaml.load(file("your.yml"))'  'cPickle.dumps(y)'
````

Time deserialization:

````
python -m timeit -s 'import cPickle, yaml; y = yaml.load(file("your.yml")); c = cPickle.dumps(y)'  'cPickle.loads(c)'
````

**Comment by mlaurense** [comment]
Yeah I know! It's because of the SSL certificates, they take up quite some space in the file.

Network latency is only 3ms (machines are on the same network).
There are 144 steps in the playbook. There are no real super-slow steps, everything flies over the screen when looking at the output (thanks to Mitogen!). The only noticable slow step is the generation of nginx config files using templates. These contain quite a few ifs and variables. As these steps are repeated for all domains, we also use quite some with_*; doing a quick search in de roles directory reveals 67 with_items, 1 with_together, 2 with_indexed_items, 9 with_nested.

I've removed all the '!vault' specifications from the playbook (leaving the encrypted data as a multiline variable), as the command failed with the error:
`ConstructorError: could not determine a constructor for the tag '!vault'`

Time parsing:
10 loops, best of 3: 495 msec per loop

Time serialization:
1000 loops, best of 3: 1.88 msec per loop

Time deserialization:
1000 loops, best of 3: 1.05 msec per loop

Hope this is useful for you.

**Comment by dw** [comment]
Building an intuition for overhead, generally big strings (like SSL certs!) are actually better than many tiny strings. Most of the pain in serialization are things like randomized memory access patterns (chasing around all over RAM for tiny objects in a way the CPU can't predict) and allocator usage (which often entails taking locks and triggering garbage collections)

Actually here, the serialization overhead is minimal. The parsing time is ridiculous, though, but I think Ansible only does that once, so it is not a problem.

Without having access to your playbook I can't really figure out where you're losing time. You already know it might be template-related, that's a good start. I'm not sure how much can be done (from an extension) to help templating speed, though.

One thing you could try is a run with "-vvv" on the command line, then grep for lines containing "ansible_mitogen.connection: Call". This times network latency (which is tiny in your case) and pure time spent waiting on the remote. While you probably won't be able to share those logs due to the argument lists they contain, it might help you figure out whether the slowness is on the target side or on the controller side. If it's templates, sadly I think this means controller. But if you see something surprising in there, like writing a tiny file taking 1 second, please leave a comment here :)

**Comment by dw** [comment]
Also finally, your case is really interesting, simply due to the size of the vars being shipped to the target. Would you mind giving me a before/after output of /usr/bin/time  with and without the plugin? Thanks!

**Comment by mlaurense** [comment]
Oh, performance improvement using Mitogen is *huge*. As mentioned before, running with Mitogen enables takes 7m36 (give or take a few seconds). Without Mitogen, the same run takes 19m49! I'm not even deploying without Mitogen anymore :)

I tried running with -vvv (even -vvvv), but this doesn't output any lines containing 'connection' or ': Call'. Is there anything else I must enable?

**Comment by dw** [comment]
Aah, you might need to grab the latest extension -- the logging integration is less than 24 hours old, so still fixing things up :)

**Comment by dw** [comment]
Thanks again so much -- just one final curiosity, how many times per day do you deploy on average? All the best

**Comment by mlaurense** [comment]
Ok, I will test again. Should have known by now, after having already done some git pulls :)

We don't deploy very much just yet (we manage about 15 server with ansible), although we are actively testing at the moment. We're integrating ansible into our webbased portal to manage all the servers, and deploy from there. This requires some testing, so we're doing more deployments than normal. As a test, we use the configuration from the server hosting the most sites, ergo, the large Yaml file.

**Comment by mlaurense** [comment]
I did see one call (of a total of 2532) taking more than one second, but I think this could also be caused by the system doing something else (e.g. a cron job).

I am going to refactor part of my playbook. The templating is apparently taking a long time. Problem is that I'm doing this step twice: I template all nginx configs (one per website), then request LetsEncrypt certificates for websites that have it enabled, then run the nginx template step again (to enable SSL). I'm probably better off using the output of the LetsEncrypt task to only re-template the websites that actually changed and make ansible skip the step for the websites that don't need reconfiguring, where as it currently builds the template (which takes time), checks it, and reports it as 'ok'.

**Comment by mlaurense** [comment]
I refactored my playbook, and now only run the second nginx templating step when necessary, using the output of the LetsEncrypt task instead of all sites. This saves quite some time, as ansible doesn't have to re-template 174 nginx config files. This shaves off another minute or so, bringing the time to run the playbook (with the SSL certificate playbook separated) to 4m26s.

**Comment by dw** [comment]
Hi @mlaurense,

It's a little late now, but it's possible Ansible does not in fact parse your huge file once. In that case, ensure that PyYAML is installed with the C extension compiled -- this is a pain, since it silently installs a very slow pure-Python version if anything is wrong. Just ensuring a native YAML parser is available should immensely help your configuration.
